File: sdk/typescript/src/intent.ts
import { Keypair, PublicKey } from '@solana/web3.js';
import bs58 from 'bs58';
import nacl from 'tweetnacl';
import { randomBytes } from 'crypto';

import {
  bytesToHex,
  canonicalSigningPayload,
  hashSigningPayload,
} from './canonical';
import {
  AccountMeta,
  ActionType,
  CanonicalAccountMeta,
  CanonicalIntent,
  GlyphError,
  IntentConstraints,
  TransactionIntent,
} from './types';

const MAX_SLIPPAGE_BPS = 10_000;
const MAX_ACCOUNTS = 256;
const MAX_U64 = 0xffffffffffffffffn;

function decodeBase58Pubkey(value: string): Uint8Array {
  const decoded = bs58.decode(value);
  const bytes = decoded instanceof Uint8Array ? decoded : Uint8Array.from(decoded);
  if (bytes.length !== 32) {
    throw new GlyphError(
      'SerializationError',
      `pubkey ${value} decoded to ${bytes.length} bytes, expected 32`,
    );
  }
  return bytes;
}

function decodeBase64(value: string): Uint8Array {
  // Buffer is faster and avoids node-vs-browser atob differences here; the SDK
  // is Node-only at the moment.
  return new Uint8Array(Buffer.from(value, 'base64'));
}

function toCanonicalAccount(account: AccountMeta): CanonicalAccountMeta {
  return {
    pubkey: decodeBase58Pubkey(account.pubkey),
    is_signer: account.is_signer,
    is_writable: account.is_writable,
  };
}

export interface BuildOptions {
  /** Override the timestamp embedded in the canonical payload. Defaults to wall clock. */
  timestamp?: bigint;
}

export class IntentBuilder {
  private constraints: Partial<IntentConstraints> = {};

  private expiryWindowSeconds = 300;

  private nonceOverride?: Uint8Array;

  private policyCommitment: Uint8Array = new Uint8Array(32);

  private workerPubkey?: Uint8Array;

  private epoch = 0n;

  // ─── Action setters ─────────────────────────────────────────────────────────
  //
  // Setters are independent — none silently fills in defaults for the other
  // fields. `build()` requires every field to be explicitly set.

  private actionTypeValue?: ActionType;

  private targetProgramValue?: string;

  private accountsValue: AccountMeta[] = [];

  private dataValue?: string;

  actionType(type: ActionType): IntentBuilder {
    this.actionTypeValue = type;
    return this;
  }

  targetProgram(program: string): IntentBuilder {
    this.targetProgramValue = program;
    return this;
  }

  accounts(accounts: AccountMeta[]): IntentBuilder {
    if (accounts.length > MAX_ACCOUNTS) {
      throw new GlyphError(
        'SerializationError',
        `accounts.length=${accounts.length} exceeds DSL Rule 7 cap of ${MAX_ACCOUNTS}`,
      );
    }
    this.accountsValue = accounts;
    return this;
  }

  data(base64Data: string): IntentBuilder {
    this.dataValue = base64Data;
    return this;
  }

  // ─── Constraint setters ────────────────────────────────────────────────────

  maxLamports(maxLamports: bigint): IntentBuilder {
    if (typeof maxLamports !== 'bigint') {
      throw new GlyphError(
        'SerializationError',
        'maxLamports must be a bigint (u64)',
      );
    }
    if (maxLamports < 0n || maxLamports > MAX_U64) {
      throw new GlyphError(
        'SerializationError',
        `maxLamports out of u64 range: ${maxLamports.toString()}`,
      );
    }
    this.constraints.maxLamports = maxLamports;
    return this;
  }

  maxSlippageBps(maxSlippageBps: number): IntentBuilder {
    if (
      !Number.isInteger(maxSlippageBps) ||
      maxSlippageBps < 0 ||
      maxSlippageBps > MAX_SLIPPAGE_BPS
    ) {
      throw new GlyphError(
        'SerializationError',
        `maxSlippageBps must be an integer in 0..=${MAX_SLIPPAGE_BPS}`,
      );
    }
    this.constraints.maxSlippageBps = maxSlippageBps;
    return this;
  }

  allowedTokens(allowedTokens: string[]): IntentBuilder {
    this.constraints.allowedTokens = allowedTokens;
    return this;
  }

  expirySeconds(seconds: number): IntentBuilder {
    if (!Number.isFinite(seconds) || seconds <= 0) {
      throw new GlyphError(
        'SerializationError',
        'expirySeconds must be a positive number',
      );
    }
    this.expiryWindowSeconds = seconds;
    return this;
  }

  /**
   * Use a caller-supplied 32-byte nonce instead of generating one. Closes F-11
   * — lets callers dedup on the client side and write deterministic replay tests.
   */
  withNonce(nonce: Uint8Array): IntentBuilder {
    if (nonce.length !== 32) {
      throw new GlyphError(
        'SerializationError',
        `nonce must be 32 bytes, got ${nonce.length}`,
      );
    }
    this.nonceOverride = nonce;
    return this;
  }

  policyCommitmentBytes(commitment: Uint8Array): IntentBuilder {
    if (commitment.length !== 32) {
      throw new GlyphError(
        'SerializationError',
        'policy_commitment must be 32 bytes',
      );
    }
    this.policyCommitment = commitment;
    return this;
  }

  workerPubkeyBytes(pubkey: Uint8Array | undefined): IntentBuilder {
    if (pubkey === undefined) {
      this.workerPubkey = undefined;
      return this;
    }
    if (pubkey.length !== 32) {
      throw new GlyphError(
        'SerializationError',
        'worker_pubkey must be 32 bytes',
      );
    }
    this.workerPubkey = pubkey;
    return this;
  }

  setEpoch(epoch: bigint | number): IntentBuilder {
    const value = typeof epoch === 'bigint' ? epoch : BigInt(epoch);
    if (value < 0n || value > MAX_U64) {
      throw new GlyphError(
        'SerializationError',
        `epoch out of u64 range: ${value.toString()}`,
      );
    }
    this.epoch = value;
    return this;
  }

  // ─── Build ─────────────────────────────────────────────────────────────────

  /**
   * Build the canonical intent struct (without signing). Useful when you want
   * to compute the signing-payload hash or pre-flight without producing a
   * wire-format payload.
   */
  buildCanonical(agent: PublicKey, options: BuildOptions = {}): CanonicalIntent {
    if (!this.actionTypeValue) {
      throw new GlyphError('SerializationError', 'Intent action type is required');
    }
    if (!this.targetProgramValue) {
      throw new GlyphError(
        'SerializationError',
        'Intent target program is required',
      );
    }
    if (!this.dataValue) {
      throw new GlyphError('SerializationError', 'Intent data (base64) is required');
    }
    if (this.constraints.maxLamports === undefined) {
      throw new GlyphError(
        'SerializationError',
        'Intent maxLamports constraint is required',
      );
    }

    const accounts = this.accountsValue.map(toCanonicalAccount);
    if (accounts.length > MAX_ACCOUNTS) {
      throw new GlyphError(
        'SerializationError',
        `accounts.length=${accounts.length} exceeds DSL Rule 7 cap`,
      );
    }

    const timestamp =
      options.timestamp ?? BigInt(Math.floor(Date.now() / 1000));
    const expiry = timestamp + BigInt(this.expiryWindowSeconds);
    const nonce = this.nonceOverride ?? new Uint8Array(randomBytes(32));

    const allowedTokens =
      this.constraints.allowedTokens === undefined
        ? undefined
        : this.constraints.allowedTokens.map(decodeBase58Pubkey);

    const canonical: CanonicalIntent = {
      agent_pubkey: agent.toBytes(),
      nonce,
      target_program: decodeBase58Pubkey(this.targetProgramValue),
      accounts,
      data: decodeBase64(this.dataValue),
      max_lamports: this.constraints.maxLamports,
      max_slippage_bps: this.constraints.maxSlippageBps,
      allowed_tokens: allowedTokens,
      expiry,
      timestamp,
      policy_commitment: this.policyCommitment,
      worker_pubkey: this.workerPubkey,
      epoch: this.epoch,
    };

    return canonical;
  }

  /**
   * Build, sign, and produce a wire-format `TransactionIntent`. The signature
   * covers `hashSigningPayload(canonicalIntent)` (SHA-256 of the canonical
   * binary bytes) — byte-for-byte identical to what the Rust SDK and TEE
   * worker compute (closes F-1, F-2).
   */
  build(keypair: Keypair, options: BuildOptions = {}): TransactionIntent {
    const canonical = this.buildCanonical(keypair.publicKey, options);
    const digest = hashSigningPayload(canonical);
    const signature = nacl.sign.detached(digest, keypair.secretKey);

    const wire: TransactionIntent = {
      version: 1,
      agent_pubkey: keypair.publicKey.toBase58(),
      nonce: bytesToHex(canonical.nonce),
      timestamp: Number(canonical.timestamp),
      expiry: Number(canonical.expiry),
      action: {
        type: this.actionTypeValue!,
        target_program: this.targetProgramValue!,
        accounts: this.accountsValue,
        data: this.dataValue!,
      },
      constraints: {
        max_lamports: this.constraints.maxLamports!.toString(),
        ...(this.constraints.maxSlippageBps !== undefined
          ? { max_slippage_bps: this.constraints.maxSlippageBps }
          : {}),
        ...(this.constraints.allowedTokens !== undefined
          ? { allowed_tokens: this.constraints.allowedTokens }
          : {}),
      },
      policy_commitment: bytesToHex(canonical.policy_commitment),
      ...(canonical.worker_pubkey !== undefined
        ? { worker_pubkey: bytesToHex(canonical.worker_pubkey) }
        : {}),
      epoch: Number(canonical.epoch),
      signature: Buffer.from(signature).toString('base64'),
    };

    return wire;
  }
}

File: sdk/rust/src/intent.rs
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use base64::Engine;
use solana_sdk::pubkey::Pubkey;
use solana_sdk::signature::Signer;
use solana_sdk::signer::keypair::Keypair;

use glyph_common::{
    canonical_signing_payload, CanonicalAccountMeta, CanonicalIntent,
};

use crate::errors::GlyphSdkError;
use crate::types::{AccountMeta, IntentAction, IntentConstraints, TransactionIntent};

#[derive(Debug, Clone, Copy)]
pub enum ActionType {
    Swap,
    Transfer,
    Stake,
    CpiCall,
}

impl ActionType {
    fn as_str(self) -> &'static str {
        match self {
            Self::Swap => "swap",
            Self::Transfer => "transfer",
            Self::Stake => "stake",
            Self::CpiCall => "cpi_call",
        }
    }
}

#[derive(Debug, Clone)]
pub struct IntentBuilder {
    version: u8,
    action_type: Option<ActionType>,
    target_program: Option<String>,
    accounts: Vec<AccountMeta>,
    data: String,
    max_lamports: Option<u64>,
    max_slippage_bps: Option<u16>,
    allowed_tokens: Option<Vec<String>>,
    ttl: Duration,
    nonce: Option<[u8; 32]>,
    policy_commitment: [u8; 32],
    worker_pubkey: Option<[u8; 32]>,
    epoch: u64,
}

impl Default for IntentBuilder {
    fn default() -> Self {
        Self {
            version: 1,
            action_type: None,
            target_program: None,
            accounts: Vec::new(),
            data: String::new(),
            max_lamports: None,
            max_slippage_bps: None,
            allowed_tokens: None,
            ttl: Duration::from_secs(60),
            nonce: None,
            policy_commitment: [0u8; 32],
            worker_pubkey: None,
            epoch: 0,
        }
    }
}

impl IntentBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn version(mut self, version: u8) -> Self {
        self.version = version;
        self
    }

    pub fn action_type(mut self, action_type: ActionType) -> Self {
        self.action_type = Some(action_type);
        self
    }

    pub fn target_program(mut self, target_program: impl Into<String>) -> Self {
        self.target_program = Some(target_program.into());
        self
    }

    pub fn account(mut self, account: AccountMeta) -> Self {
        self.accounts.push(account);
        self
    }

    pub fn accounts(mut self, accounts: impl IntoIterator<Item = AccountMeta>) -> Self {
        self.accounts.extend(accounts);
        self
    }

    pub fn data(mut self, data: impl Into<String>) -> Self {
        self.data = data.into();
        self
    }

    pub fn max_lamports(mut self, max_lamports: u64) -> Self {
        self.max_lamports = Some(max_lamports);
        self
    }

    pub fn max_slippage_bps(mut self, max_slippage_bps: u16) -> Self {
        self.max_slippage_bps = Some(max_slippage_bps);
        self
    }

    pub fn allowed_tokens(mut self, allowed_tokens: Vec<String>) -> Self {
        self.allowed_tokens = Some(allowed_tokens);
        self
    }

    pub fn ttl(mut self, ttl: Duration) -> Self {
        self.ttl = ttl;
        self
    }

    /// Override the random nonce — useful for tests and replay-recovery.
    /// Closes AUDIT_SDK F-11.
    pub fn with_nonce(mut self, nonce: [u8; 32]) -> Self {
        self.nonce = Some(nonce);
        self
    }

    /// Bind the policy commitment that the worker is expected to be enforcing.
    /// Without this, a delegation could be replayed against a worker running
    /// a different policy. Required for parity with the worker-side check.
    pub fn policy_commitment(mut self, policy_commitment: [u8; 32]) -> Self {
        self.policy_commitment = policy_commitment;
        self
    }

    /// Bind the worker's Ed25519 public key the agent expects to be talking to.
    pub fn worker_pubkey(mut self, worker_pubkey: [u8; 32]) -> Self {
        self.worker_pubkey = Some(worker_pubkey);
        self
    }

    pub fn epoch(mut self, epoch: u64) -> Self {
        self.epoch = epoch;
        self
    }

    pub fn build(self, agent_keypair: &Keypair) -> Result<TransactionIntent, GlyphSdkError> {
        let action_type = self.action_type.ok_or_else(|| {
            GlyphSdkError::IntentBuildError("action_type is required".to_string())
        })?;

        let target_program = self.target_program.ok_or_else(|| {
            GlyphSdkError::IntentBuildError("target_program is required".to_string())
        })?;

        let target_program_pk = target_program.parse::<Pubkey>().map_err(|_| {
            GlyphSdkError::IntentBuildError(
                "target_program must be a valid Solana pubkey".to_string(),
            )
        })?;

        let max_lamports = self.max_lamports.ok_or_else(|| {
            GlyphSdkError::IntentBuildError("max_lamports is required".to_string())
        })?;

        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_err(|e| GlyphSdkError::IntentBuildError(format!("invalid system clock: {e}")))?
            .as_secs();

        let expiry = now
            .checked_add(self.ttl.as_secs())
            .ok_or_else(|| GlyphSdkError::IntentBuildError("expiry overflow".to_string()))?;

        let nonce_bytes: [u8; 32] = match self.nonce {
            Some(n) => n,
            None => Keypair::new().pubkey().to_bytes(),
        };
        let nonce_hex = hex::encode(nonce_bytes);

        // Decode account pubkeys for the canonical payload, while preserving
        // the wire-format string accounts in the intent itself.
        let mut canonical_accounts = Vec::with_capacity(self.accounts.len());
        for meta in &self.accounts {
            let pk = meta.pubkey.parse::<Pubkey>().map_err(|_| {
                GlyphSdkError::IntentBuildError(format!(
                    "invalid account pubkey: {}",
                    meta.pubkey
                ))
            })?;
            canonical_accounts.push(CanonicalAccountMeta {
                pubkey: pk.to_bytes(),
                is_signer: meta.is_signer,
                is_writable: meta.is_writable,
            });
        }

        let data_bytes = base64::engine::general_purpose::STANDARD
            .decode(&self.data)
            .map_err(|e| {
                GlyphSdkError::IntentBuildError(format!("data must be base64: {e}"))
            })?;

        let allowed_tokens_bytes = match &self.allowed_tokens {
            Some(tokens) => {
                let mut out = Vec::with_capacity(tokens.len());
                for t in tokens {
                    let pk = t.parse::<Pubkey>().map_err(|_| {
                        GlyphSdkError::IntentBuildError(format!(
                            "allowed_token is not a valid pubkey: {t}"
                        ))
                    })?;
                    out.push(pk.to_bytes());
                }
                Some(out)
            }
            None => None,
        };

        let action = IntentAction {
            action_type: action_type.as_str().to_string(),
            target_program,
            accounts: self.accounts,
            data: self.data,
        };

        let constraints = IntentConstraints {
            max_lamports,
            max_slippage_bps: self.max_slippage_bps,
            allowed_tokens: self.allowed_tokens,
        };

        let canonical = CanonicalIntent {
            agent_pubkey: agent_keypair.pubkey().to_bytes(),
            nonce: nonce_bytes,
            target_program: target_program_pk.to_bytes(),
            accounts: canonical_accounts,
            data: data_bytes,
            max_lamports,
            max_slippage_bps: self.max_slippage_bps,
            allowed_tokens: allowed_tokens_bytes,
            expiry,
            timestamp: now,
            policy_commitment: self.policy_commitment,
            worker_pubkey: self.worker_pubkey,
            epoch: self.epoch,
        };

        let payload = canonical_signing_payload(&canonical);
        let signature = agent_keypair.sign_message(&payload);

        let intent = TransactionIntent {
            version: self.version,
            agent_pubkey: agent_keypair.pubkey().to_string(),
            nonce: nonce_hex,
            timestamp: now,
            expiry,
            action,
            constraints,
            signature: base64::engine::general_purpose::STANDARD.encode(signature.as_ref()),
        };

        Ok(intent)
    }
}
